cpp/src/arrow/dataset/filter.cc

```diff
@@ -1337,7 +1351,11 @@ Result<std::shared_ptr<RecordBatch>> TreeEvaluator::Filter(
   return batch->Slice(0, 0);
 }
 
-std::shared_ptr<Expression> scalar(bool value) { return scalar(MakeScalar(value)); }
+const std::shared_ptr<Expression>& scalar(bool value) {
+  static auto true_ = scalar(MakeScalar(true));
+  static auto false_ = scalar(MakeScalar(false));
+  return value ? true_ : false_;
+}
```

Comment:
Just results in slightly less allocator churn for the very common case of a boolean ScalarExpression, since we produce only two permanent instances rather than one per invocation of `scalar(bool)`